Enforce a style guide

magopian opened the pull request:
Fixes #77

To run eslint:
- for node: `npm run lint-node` (use `npm run lint-node-fix` to auto fix some issues)
- for the browser: `npm run lint-browser` (use `npm run lint-browser-fix` to auto fix some issues)
- for both: `npm run lint` (use `npm run lint-fix` to auto fix some issues)

magopian commented:
Here's a run for the node files: https://dpaste.de/cay1

jankeromnes commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,69 @@
+  "extends": [
+    "eslint:recommended",
+    "plugin:node/recommended",
+    "standard"
```
Interesting, what are these `eslint:recommended` and `plugin:node/recommended` rules? Also, are they not already included in `standard`?

magopian replied:
I believe those are base rules that are "cascading"? Here's the `eslint:recommended`: http://eslint.org/docs/rules/

I'm not sure which is where, I just feel it's sensible to do it like that, but we could very well just keep `standard` I guess?

jankeromnes commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,69 @@
+      "experimentalObjectRestSpread": true,
+      "jsx": true
+    },
+    "sourceType": "module"
```
I don't get these parser options for the browser environment. Since we don't use ES6, why add `"ecmaVersion": "2017"`? Also, I don't believe we do any experimental object rest spread or JSX here (which is React's non-standard JS format).

jankeromnes commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,69 @@
+    "node"
+  ],
+  "rules": {
+    // Standard js rules
```
Nit: These are actually rules that contradict standard.

jankeromnes commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,69 @@
+  "rules": {
+    // Standard js rules
+    "semi": ["error", "always"],
+    "comma-dangle": ["error", "never"],
```
Nit: We try to use dangling commas on multiline objects and arrays. Please use something like `"comma-dangle": ["error", "only-multiline"]` (or you may want to explicitly list all [options](http://eslint.org/docs/rules/comma-dangle#options) in order to special-case `"functions": "never"`, because it may break older browsers).

magopian replied:
not in the browser, as discussed on irc. Or did I misunderstand?

jankeromnes commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,69 @@
+    "comma-dangle": ["error", "never"],
+    
+    // Node require rules
+    "node/exports-style": ["error", "module.exports"],
```
We don't use `exports` or `module.exports` in the browser.

jankeromnes commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,69 @@
+    // overriding recommended rules
+    "no-constant-condition": ["error", { checkLoops: false }],
+    "no-console": [ "error", { allow: ["log", "warn", "error"] } ],
+    "no-unused-vars": ["error", { "argsIgnorePattern": "^_" }],
```
Interesting, why did you add these?

magopian replied:
mmmm, copy pasted from the html profiler devtools, i'll remove them.

jankeromnes commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,69 @@
+    "prefer-const": "error",
+    "prefer-promise-reject-errors": "error",
+    "prefer-rest-params": "error",
+    "prefer-spread": "error",
```
Why did you add all these? Aren't they examples already present in `"standard"`? Also a few of them are ES6+ and don't make sense in the browser.

magopian replied:
same as above

jankeromnes commented on .eslintrc-node.js:
```diff
@@ -0,0 +1,78 @@
+module.exports = {
+  "env": {
+    "browser": true,
```
Why set `"browser": true` in server files?

jankeromnes commented on package.json:
```diff
@@ -37,7 +37,15 @@
-  "engines" : {
-    "node" : ">=6.0.0"
+  "engines": {
+    "node": ">=6.0.0"
```
Thank you for fixing this bogus whitespace!

magopian commented:
I added some automatic git hooks that will run the linting before each commit and push. You'll need to run `npm install` to activate them.

When commiting or pushing you should see something like this (check the little spinner next the "lint-browser-fix" line):

<img width="682" alt="screen shot 2017-07-07 at 16 55 54" src="https://user-images.githubusercontent.com/167767/27963763-431cd6f6-6336-11e7-9e45-2efe5530f704.png">

If the checks fail, you'll be prompted with the following message:

```husky > pre-commit hook failed (add --no-verify to bypass)```

So if you wish to bypass the checks, simply use `git commit --no-verify ....`

jankeromnes commented on lib/configurations.js:
```diff
@@ -10,23 +10,23 @@ const db = require('./db');
+  '.nanorc',
+  '.netrc',
+  '.ssh/config',
+  '.vimrc',
```
Heh, @nt1m is not going to like this 😛

nt1m replied:
Noooooooooo

nt1m commented on static/js/bootstrap-3.3.7.min.js:
```diff
@@ -3,5 +3,8 @@
  * Copyright 2011-2016 Twitter, Inc.
  * Licensed under the MIT license
  */
-if("undefined"==typeof jQuery)throw new Error("Bootstrap's JavaScript requires jQuery");+function(a){"use strict";var b=a.fn.jquery.split(" ")[0].split(".");if(b[0]<2&&b[1]<9||1==b[0]&&9==b[1]&&b[2]<1||b[0]>3)throw new Error("Bootstrap's JavaScript requires jQuery version 1.9.1 or higher, but lower than version 4")}(jQuery),+function(a){"use strict";function b(){var a=document.createElement("boot ...
```
Can you leave libs unchanged and add a .eslintignore file including common library patterns ?

magopian replied:
What kind of "common library patterns"? Keep in mind I barely know any JS ;)

nt1m replied:
static/js/*.min.js as you figured out :)

magopian replied:
ah, that, sorry, didn't understand your statement at first ;)

magopian commented:
Added an `.eslintignore` file, and ignored all files ending with `.min.js`. There aren't many errors left, and the linting step before commiting/pushing is not really short (less than a second on my machine), which is imho totally acceptable.

jankeromnes commented on .eslintrc-node.js:
```diff
@@ -0,0 +1,28 @@
+        "objects": "only-multiline",
+        "imports": "never",
+        "exports": "never",
+        "functions": "never",
```
Nit: These rules should be in the browser rules, not in the Node.js/server rules. In the node rules, it should be `'command-dangle': ['error', 'only-multiline']`.

magopian replied:
iiuc the browsers are not all on board with allowing dangling commas, that's why i have https://github.com/JanitorTechnology/janitor/pull/101/files#diff-a75257962af8dc9ccbfdb1d2c43be2feR13 which disables those checks. But then maybe I should rewrite this to `"comma-dangle": "off"` instead, it would be clearer I guess.

Regarding node, it does allow dangling commas, right? So that's why I have `"only-multiline"`, but as we discussed on irc, only for objects and arrays, nothing else, or did I misunderstand? So appart from rewriting the browser rule to be clearly `"off"`, what else do you want me to modify?

jankeromnes commented on api/hosts-api.js:
```diff
@@ -149,7 +149,6 @@ hostAPI.post({
     // No authentication.
     response.statusCode = 403; // Forbidden
     response.json({ error: 'Unauthorized' }, null, 2);
-    return;
```
Interesting. This `return` was indeed unnecessary, but I added it as an indication that there should be no further instructions in this block (below, there are some inline functions, but there shouldn't be any other code or logic that risks calling `response.json()` or `response.end()` a second time). This is just a random remark, I'm fine with this change.

jankeromnes commented on api/user-api.js:
```diff
@@ -154,7 +154,7 @@ configurationsAPI.patch({
     },
     response: {
-      body: JSON.stringify({ '.gitconfig': '[user]\nname = Janitor'  }, null, 2)
+      body: JSON.stringify({ '.gitconfig': '[user]\nname = Janitor' }, null, 2)
```
Nice catch! 😄

magopian replied:
It was an auto-fix <3

jankeromnes commented on lib/certificates.js:
```diff
@@ -485,5 +446,4 @@ function requestLetsEncryptCertificate (parameters, callback) {
   };
 
   client.getCertificate(options, callback);
-
```
Huge 👍! I used to do all these changes by hand... You're saving me so much time!

magopian replied:
Same here, just an auto-fix

jankeromnes commented on package.json:
```diff
@@ -22,7 +22,13 @@
+    "lint-browser": "eslint -c .eslintrc-browser.js static/",
+    "lint-browser-fix": "npm run lint-browser -- --fix",
+    "precommit": "lint-staged",
+    "prepush": "lint-staged"
```
Actually, I'm not fond of commit hooks, and @nt1m is worried about them too. Do you mind removing them from this pull request, and maybe suggest them in a follow-up pull request if you really think they're really valuable?

jankeromnes commented on static/js/graphs.js:
```diff
@@ -23,16 +23,13 @@ Array.map(document.querySelectorAll('*[data-data]'), function (div) {
       }
     },
-    labelsUTC: true,
+    labelsUTC: true
```
Nit: I believe we should allow dangling commas for multi-line objects and arrays in browser code.

magopian replied:
Wow, so I did completely misunderstand our discussion on IRC. My understanding was that browsers don't all allow them, but node does. Can you please confirm that it's the opposite?

jankeromnes replied:
So, most browsers support:

```js
var a = {
  b: 4,
  c: 5,
};
```

And:

```js
var a = [
  c,
  d,
];
```

But only very recent browsers support:

```js
function bla(
  a,
  b,
  c,
) { }
```

So in browser code, we want to forbid dangling spaces in function arguments (as well as in imports/exports since they don't make sense in a browser env anyway).

In Node.js code, we can add dangling commas everywhere (i.e. on every multi-line thing), even function parameters.

magopian replied:
I believe I changed it to do just that, if I'm wrong, please let me know ;)

jankeromnes commented on static/js/graphs.js:
```diff
@@ -50,23 +47,4 @@ function formatTime (milliseconds) {
-    p++;
-  }
-
-  return (Math.round(value * 100) / 100) + ' ' + prefix[p] + 'B';
```
Please don't delete this function! I know it's not used yet, but we might want to use it in the future.

magopian replied:
Ok I'll revert this commit ;)

jankeromnes commented on static/js/janitor.js:
```diff
@@ -99,7 +101,7 @@ function fetchAPI (method, url, data, callback) {
     // The server is responding!
     responseStatus = response.status;
-    return responseStatus == 204 ? null : response.json();
+    return responseStatus === 204 ? null : response.json();
```
👍!

nt1m commented on static/js/settings.js:
```diff
@@ -1,11 +1,11 @@
 // Copyright © 2015 Jan Keromnes. All rights reserved.
 // The following code is covered by the AGPL-3.0 license.
 
+/* global ajaxForm, updateFormStatus */
```
I think you can use something like import-globals-from or define something in static/.eslintrc.js

magopian commented:
i believe I've addressed all your comments (but https://github.com/JanitorTechnology/janitor/pull/101#discussion_r126185661, should I do something about that?).

Please let me know if there's anything else you wish me to modify ;)

jankeromnes commented on join.js:
```diff
@@ -16,9 +16,10 @@ const sessions = require('./lib/sessions');
+  const error = '[fail] cannot join cluster as [hostname = ' + hostname + ']: ' +
+    'please fix the hostname in ./db.json and try again';
+  log(error);
+  throw new Error(error);
```
Nit: Please either `log('[fail] something something')` or `throw new Error('Something something')`, but not both. (In this case, I believe you want to just `throw`, and thus there is no need for the `[fail]` log prefix).

jankeromnes commented on join.js:
```diff
@@ -258,16 +259,16 @@ function routeRequest (proxyParameters, request, response) {
       const url = 'https://' + hostname + ':' + port + path;
       routes.redirect(response, url);
       return;
+    }
```
Nit: What? Why?

magopian replied:
It's my understanding that those are needed to fix http://eslint.org/docs/rules/no-case-declarations

nt1m replied:
Can you simply move the contents of `url`  inside routes.redirect and remove the extra variable and the brackets?

jankeromnes commented on lib/boot.js:
```diff
@@ -263,7 +263,7 @@ exports.ensureDockerTlsCertificates = function (next) {
           }
 
-          fs.chmod(path, 0600 /* read + write by owner */, (error) => {
+          fs.chmod(path, '0600' /* read + write by owner */, (error) => {
```
Why change this to a String? The Node.js [fs.chmod documentation](https://nodejs.org/api/fs.html#fs_fs_chmod_path_mode_callback) says that the `mode` argument should be an Integer (even though it does unofficially support strings). There is no example, but the default value is said to be `0o666` in [other places](https://nodejs.org/api/fs.html#fs_fs_appendfile_file_data_options_callback).

magopian replied:
The octal syntax seems to be deprecated, see http://eslint.org/docs/rules/no-octal (first paragraph). In strict mode, eslint will complain about a parsing error.

nt1m replied:
I prefer that we use `0o666` rather than `'0600'`.

jankeromnes commented on lib/certificates.js:
```diff
@@ -1,7 +1,7 @@
 // The following code is covered by the AGPL-3.0 license.
 
-let child_process = require('child_process');
+let childProcess = require('child_process');
```
Grmbl... fair enough. But we might as well make these const, by enabling the [prefer-const](http://eslint.org/docs/rules/prefer-const) ESLint rule.

jankeromnes commented on lib/docker.js:
```diff
@@ -1,14 +1,15 @@
 // The following code is covered by the AGPL-3.0 license.
 
-const dockerode = require('dockerode');
+const Dockerode = require('dockerode');
```
Nit: What? Why?

Edit: Hm, because of the usage with `new`? ... Well, ok then, if it must be.

jankeromnes commented on lib/docker.js:
```diff
@@ -206,12 +207,14 @@ exports.removeContainer = function (parameters, callback) {
 
   getDocker(host, (error, docker) => {
+    if (error) {
+      log('[fail] Error while getting docker', error);
```
Nit: In general, log messages are lower-case, while error messages in `new Error()`s are capitalized.

Nit 2: Please add the word `client` as in `while getting docker client`.

jankeromnes commented on lib/hosts.js:
```diff
@@ -33,8 +33,8 @@ exports.get = function (hostname) {
 exports.authenticate = function (request) {
-  let { client_id = null, client_secret = null } = request.query;
-  if (!client_id || !client_secret) {
+  let { clientId = null, clientSecret = null } = request.query;
```
⚠️ This change breaks the destructuring (`request.query` specifically contains the keys `client_id` and `client_secret`, but not `clientId` nor `clientSecret`).

Hint: The correct fix would be:

```js
let { client_id: clientId = null, client_secret: clientSecret = null } = request.query;
```

But that's very long and unreadable. I'll let you decide whether to continue using `client_id` everywhere, or fix the destructuring. (Maybe it might become more readable multi-line?)

Hint 2: @nt1m suggested not using destructuring in this case, i.e.:

```js
let clientId = request.query.client_id || null;
let clientSecret = request.query.client_secret || null;
```

magopian replied:
I'm reverting all those changes and allowing non-camelcase names instead. Thanks for spotting this!

jankeromnes commented on lib/hosts.js:
```diff
@@ -155,17 +155,17 @@ exports.resetOAuth2ClientSecret = function (host, callback) {
 exports.issueOAuth2AuthorizationCode = function (request, callback) {
-  let { client_id, scope, state, redirect_url = null } = request.query;
-  const hostname = exports.identify(client_id);
+  let { clientId, scope, state, redirectUrl = null } = request.query;
```
⚠️ Same as above, broken destructuring.

jankeromnes commented on lib/hosts.js:
```diff
@@ -188,21 +188,21 @@ exports.issueOAuth2AuthorizationCode = function (request, callback) {
 exports.issueOAuth2AccessToken = function (request, callback) {
-  const { client_id, code, state } = request.query;
-  const hostname = exports.identify(client_id);
+  const { clientId, code, state } = request.query;
```
⚠️ Broken destructuring.

jankeromnes commented on lib/hosts.js:
```diff
@@ -188,21 +188,21 @@ exports.issueOAuth2AuthorizationCode = function (request, callback) {
       'code=' + encodeURIComponent(code) +
       '&state=' + encodeURIComponent(state);
-    callback(null, { code, redirect_url });
+    callback(null, { code, redirectUrl });
```
⚠️ Caution: This changes the object format sent with the callback. It will no longer include a key called `redirect_url`, and instead include `redirectUrl`. Are you sure this is ok? Otherwise, you can use `{ code, redirect_url: redirectUrl }` to preserve the output format.

jankeromnes commented on lib/machines.js:
```diff
@@ -87,7 +87,7 @@ exports.rebuild = function (projectId, callback) {
   docker.buildImage({ host, tag, dockerfile }, (error, stream) => {
     if (error) {
-      log ('rebuild', image, error);
+      log('rebuild', image, error);
```
Please drive-by fix this log message to include the prefix `[fail]`.

jankeromnes commented on lib/machines.js:
```diff
@@ -288,7 +288,7 @@ exports.destroy = function (user, projectId, machineId, callback) {
 exports.deployConfigurationInAllContainers = function (user, file) {
   let count = 0;
-  const machines = Object.values(user.machines)
+  const machines = Object.values(user.machines) // eslint-disable-line node/no-unsupported-features
```
Nit: Doesn't this line overshoot the 80-char limit? Also, we didn't have a debate about the whole 80-char vs 100-char vs no-limit story 😄

nt1m replied:
You can use:
```js
// eslint-disable-next-line node/no-unsupported-features
const machines = Object.values(user.machines)
```

jankeromnes commented on lib/oauth2.js:
```diff
@@ -106,6 +107,10 @@ exports.request = function (parameters, callback) {
 
+    if (error) {
+      log('[fail] Error while getting the client', error);
+    }
```
Nit: I prefer `error` parameters to be the first thing handled in a function. Please move it to the function's first line, i.e. above `const api = ...`.

jankeromnes replied:
Nit 2: Please use lower-case log messages (as in `error` instead of `Error`). Additionally, logging the variable `error` will already print `Error: Something something` in the console. Maybe it's not necessary to repeat the word "error" multiple times? I.e.

```
[2017-07-10T16:00:43.733Z] [fail] Error while getting the client Error: Couldn't reach host 'example.com'
```

jankeromnes commented on lib/routes.js:
```diff
@@ -364,6 +364,10 @@ exports.webProxy = function (request, response, parameters) {
 
   proxy.on('error', error => {
+    if (error) {
+      log('[fail] Error while processing the request', error);
```
Nit: Cool fix! But please use lower-case error messages (or remove the word "Error").

jankeromnes commented on lib/sessions.js:
```diff
@@ -69,6 +70,10 @@ exports.get = function (request, callback) {
 exports.destroy = function (request, callback) {
   exports.get(request, (error, session, token) => {
+    if (error) {
+      log('[fail] Error while destroying the session', error);
```
Nit: Same as earlier nits (word "Error" is superfluous).

jankeromnes commented on lib/streams.js:
```diff
@@ -3,14 +3,14 @@
 
 var stream = require('stream');
```
Whoa, this file still uses `var` 😄 Maybe it can be auto-fixed by ESLint to use `const`, or `let` if it really needs to be mutable?

magopian replied:
I will add a commit with such a rule

jankeromnes commented on lib/streams.js:
```diff
@@ -52,23 +49,24 @@ function set (object, key, readable) {
   // Remove the stream if an error occurs.
   readable.on('error', function (error) {
+    if (error) {
+      log('[fail] Error reading the stream', error);
```
Nit: Please remove the word `Error`, e.g. like so `'[fail] problem while reading stream'`.

jankeromnes commented on lib/users.js:
```diff
@@ -95,6 +95,10 @@ exports.resetSSHKeyPair = function (user) {
 exports.sendLoginEmail = function (email, request, callback) {
   sessions.get(request, (error, session, token) => {
+    if (error) {
+      log('[fail] Error sending the login email', error);
```
Nit: -Error, -the.

jankeromnes commented on lib/users.js:
```diff
@@ -145,6 +149,10 @@ exports.sendLoginEmail = function (email, request, callback) {
   // Generate a dummy token for that invite.
   sessions.create((error, token, session) => {
+    if (error) {
+      log('[fail] Error creating the session', error);
```
Nit: Maybe `'[fail] could not create session'`?

jankeromnes commented on package.json:
```diff
@@ -22,7 +22,13 @@
+    "lint-browser": "eslint -c .eslintrc-browser.js static/",
+    "lint-browser-fix": "npm run lint-browser -- --fix",
+    "lint": "npm run lint-node && npm run lint-browser",
+    "lint-fix": "npm run lint-node-fix && npm run lint-browser-fix"
```
Nit: Is it really worth it to have separate `lint-node{,-fix}` and `lint-browser{,-fix}` commands? Why not just have `lint` and `lint-fix`?

magopian replied:
You're right, I'll remove the others ;)

magopian commented:
I believe I've addressed your comments, please let me know if I've missed anything.

I added the `no-var` rule in https://github.com/JanitorTechnology/janitor/pull/101/commits/af4933dd1ed1a6aa05313a2101d83c3d331aeae0, please have a deep look at my manual changes there, to make sure I haven't set something as `const` while it should really be a `let` ;)

I also rebased the PR, so I hope I didn't break anything in the process :/

magopian commented:
Rebased following latest PR merges

nt1m commented on .eslintrc-browser.js:
```diff
@@ -0,0 +1,36 @@
+module.exports = {
+  "env": {
+    "browser": true,
+    "es6": false
```
Quite funny that the eslintrc.js files don't actually respect the eslint rules :D 

Property names should either have no quotes or single quotes.

magopian replied:
I believe it's because you can either use a js file, or plain json? I don't have an opinion on this matter. Also for some reason those configuration files are auto-ignored by eslint.

nt1m commented on join.js:
```diff
@@ -258,16 +257,16 @@ function routeRequest (proxyParameters, request, response) {
       log('[fail] unsupported proxy type:', proxy);
       response.statusCode = 500; // Internal Server Error
       response.end();
-      return;
```
It's inconsistent that we have `return;` in the above `case` statement, but not here. Can you instead use `break;` here and above to remove this inconsistency?

nt1m commented on lib/db.js:
```diff
@@ -25,38 +24,31 @@ function load () {
       console.error('Can\'t write DB file!', error.stack);
     }
-    fs.chmod(file, 0600 /* read + write by owner */, function (error) {
+    fs.chmod(file, '0600' /* read + write by owner */, function (error) {
```
Same here.

nt1m commented on lib/metrics.js:
```diff
@@ -132,23 +116,19 @@ exports.getContributionData = function () {
   let hosts = db.get('hosts');
 
-  for (let hostname in hosts) {
+  for (let hostname in hosts) { // eslint-disable-line no-unused-vars
```
Nit: please use `// eslint-disable-next-line` instead.

nt1m commented on static/js/graphs.js:
```diff
@@ -50,14 +45,11 @@ function formatTime (milliseconds) {
 
-function formatMemory (bytes) {
-
+function formatMemory (bytes) { // eslint-disable-line no-unused-vars
```
nit: I prefer you use `/* exported formatMemory */` on top of the file

magopian replied:
that doesn't seem to change eslint's mind about warning on this line :/

Probably because of https://github.com/eslint/eslint/issues/1200#issuecomment-156464573

nt1m replied:
nit: please use // eslint-disable-next-line

nt1m commented on lib/boot.js:
```diff
@@ -292,7 +292,7 @@ exports.ensureDockerTlsCertificates = function (next) {
     }
 
-    done = null;
+    done = null; // eslint-disable-line no-func-assign
```
nit: please use 
```js
// eslint-disable-next-line no-func-assign
done = null;
```

I think it's cleaner as it doesn't clutter up the line length.

nt1m commented on lib/certificates.js:
```diff
@@ -369,7 +341,7 @@ exports.createHTTPSCertificate = function (parameters, callback) {
 
     // All required tasks are now finished, destroy the `done` callback.
-    done = null;
+    done = null; // eslint-disable-line no-func-assign
```
nit: use eslint-disable-next-line

magopian commented:
@nt1m and @jankeromnes i'm not sure I've addressed all your comments, this PR is getting so long that it's hard to tell. Please let me know if I've missed anything.

On a side note, I did change the eslint exceptions to use `// eslint-disable-next-line` but I'm not sure it's a good idea: it's the kind of comments which will
- have a much bigger importance (at the same level as code comments explaining pieces of the implementation, and adding "noise", instead of just being an addition far to the right of the line
- be easily forgotten if the line that is being ignored is moved around, or fixed, or removed

nt1m commented:
> have a much bigger importance (at the same level as code comments explaining pieces of the implementation, and adding "noise", instead of just being an addition far to the right of the line

I think eslint-disable-next-line looks cleaner as it doesn't make the line too long. I do think having a line that's suddenly wider than the other ones looks odd. I would even be favorable to forbidding long inline comments.

> be easily forgotten if the line that is being ignored is moved around, or fixed, or removed

Either eslint will catch the mistake, either the comment is no longer relevant :)

jankeromnes commented on join.js:
```diff
@@ -253,21 +252,21 @@ function routeRequest (proxyParameters, request, response) {
   }
 
   const { port, proxy } = proxyParameters;
+  const url = 'https://' + hostname + ':' + port + path;
```
Nit: This `url` variable is not used by anyone else than the `routes.redirect()` below. Please inline its value directly into the redirect call (no need for an intermediary variable).

jankeromnes commented on lib/configurations.js:
```diff
@@ -70,7 +70,7 @@ exports.resetToDefault = function (user, file, callback) {
 
   let content = '';
-  stream.on('data', chunk => content += String(chunk));
+  stream.on('data', chunk => (content += String(chunk)));
```
Nit: I believe you meant `{ content += String(chunk); }` instead of `(content += String(chunk))` (we didn't actually want to return the results of the `+=`).

jankeromnes commented on package.json:
```diff
@@ -24,10 +24,12 @@
-    "watch": "watch-run --initial --pattern 'app.js,package.json,api/**,lib/**,templates/**' --stop-on-error npm run app & tailf janitor.log -n 0"
+    "watch": "watch-run --initial --pattern 'app.js,package.json,api/**,lib/**,templates/**' --stop-on-error npm run app & tailf janitor.log -n 0",
+    "lint": "eslint -c .eslintrc-node.js *.js api/ lib/ && eslint -c .eslintrc-browser.js static/",
+    "lint-fix": "eslint -c .eslintrc-node.js *.js api/ lib/ --fix && eslint -c .eslintrc-browser.js static/ --fix"
```
Nit: Please alpha-sort NPM scripts (except for `pre*` and `post*` scripts).

magopian commented on api/user-api.js:
```diff
@@ -168,7 +167,7 @@ configurationsAPI.patch({
     },
     response: {
-      body: JSON.stringify({ '.gitconfig': '[user]\nname = Sally' }, null, 2)
+      body: JSON.stringify({ '.gitconfig': '[user]\nname = Janitor' }, null, 2)
```
Yeah, I don't know where this change comes from, I believe it was from a rebase? I personally didn't make this change, do you want me to revert it?

nt1m replied:
Yeah, I think it would be preferable to revert it
